include/deal.II/base/quadrature_point_data.h

```diff
@@ -0,0 +1,615 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2016 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii__quadrature_point_data_h
+#define dealii__quadrature_point_data_h
+
+#ifdef DEAL_II_WITH_CXX11
+
+#include <deal.II/base/quadrature.h>
+#include <deal.II/base/config.h>
+#include <deal.II/base/subscriptor.h>
+#include <deal.II/grid/tria.h>
+#include <deal.II/grid/tria_accessor.h>
+#include <deal.II/distributed/tria.h>
+#include <deal.II/fe/fe_tools.h>
+#include <deal.II/fe/fe.h>
+#include <deal.II/lac/vector.h>
+
+#include <vector>
+#include <map>
+#include <typeinfo>
+
+DEAL_II_NAMESPACE_OPEN
+
+/*!@addtogroup Quadrature */
+/*@{*/
+
+/**
+ * An abstract class which specifies requirements for classes to be used as
+ * a template argument in QPManagerBase .
+ *
+ * @author Denis Davydov, Jean-Paul Pelteret, 2016
+ */
+struct QPData
+{
+public:
+  /**
+   * Constructor.
+   */
+  QPData() {};
+
+  /**
+   * Virtual distructor.
+   */
+  virtual ~QPData() {};
+
+  /**
+   * Pack all data stores in this class into a vector @p values .
+   */
+  virtual void pack_values(std::vector<double> &values) const = 0;
+
+  /**
+   * The opposite of the above, namely
+   * unpack a vector @p scalars into the data stored in this class.
+   */
+  virtual void unpack_values(const std::vector<double> &values) = 0;
+};
+
+/**
+ * Base class for storing at each cell @p CELL a vector of quadrature point (QP) data @p DATA .
+ * The underlying structure and the initialize() method of this class is designed
+ * in such a way that one could use different child class derived from the
+ * base DATA class to store QP data. This implies the usage of pointers, in our
+ * case -- std::shared_ptr().
+ *
+ * @note It is assumed that all QP's in a cell have the same construction (specifically,
+ * the number of internal data values). For this reason, this class may not be sufficiently
+ * flexible when adopting a level-set approach to describe material behavior.
+ *
+ * @author Denis Davydov, Jean-Paul Pelteret, 2016
+ */
+template <typename CELL, typename DATA>
+class QPManagerBase : public Subscriptor
```

Comment:
it is very much connected to quadrature data as `DataType` should be derived from `QuadraturePointData` and it implements some pack and unpack functions which are used in the derived class.